### DISCUSSION TITLE
Bind Table Filter Field to Javascript Variable

### DISCUSSION THREAD

**Discussion by intrepidws** [filamentphp/filament#1904] [Help] [answered]
I'm trying to implement a geolocation filter for one of my resources and running into trouble setting the location filter field to the results of the Javascript calls to get the browser location. This is the code I have so far:

```php
Forms\Components\Textinput::make('zip_code_or_long_lat')
    ->label('Zip Code')
    ->helperText('Latitude and longitude coordinates are also acceptable.')
    ->extraInputAttributes([
        'x-bind:value' => 'coords',
    ]),
```

Within the `extraInputAttributes()` method, I'm trying to bind the value of `coords` (a Javascript variable that I'm setting to the coordinates in some custom JS on the page). It _seems_ to work, since I can visibly see that the field has the coordinates - but Alpine (or Livewire) isn't aware of that value until I actually click into the field and change it.

I'm sure that I'm missing something about how the field lifecycle is supposed to work, but hoping someone might be able to point me in the right direction here.

**Comment by ryangjchandler** [comment]
Does `x-model` make a difference? Might also help if you provide some of the Alpine code too.

**Reply by intrepidws** [reply]
The Javascript variable that I'm using isn't Alpine at all, it's super basic:

```javascript
latitude = null;
longitude = null;
coords = null;

function successfulLookup(position) {
    latitude = position.coords.latitude;
    longitude = position.coords.longitude;
    coords = position.coords.latitude + ', ' + position.coords.longitude;
}

if (window.navigator.geolocation) {
    window.navigator.geolocation.getCurrentPosition(successfulLookup);
}
```

**Reply by ryangjchandler** [reply]
Ah, that'll be the problem then won't it? `x-bind:value` is Alpine specific and it'll only be binding on the first load since `coords` isn't a reactive value.

~~Where does `successfulLookup` get called?~~ Ignore me, I see it being called from the `getCurrentPosition`.

**Reply by ryangjchandler** [reply]
I think you'll want to use Alpine stores for this instead.

```js
Alpine.store('geolocation', {
	latitude: null,
	longitude: null,
	coords: null,
	init() {
		if (window.navigator.geolocation) {
			window.navigator.geolocation.getCurrentPosition(position => {
				this.latitude = position.coords.latitude;
				this.longitude = position.coords.longitude;
				this.coords = this.latitude + ', ' + this.longitude;
			})
		}
	}
})
```

Then you should be able to `'x-bind:value' => '$store.geolocation.coords'`. Haven't tested, let me know if it helps.

**Reply by intrepidws** [reply]
Thanks for your help here, I appreciate it. Unfortunately, while this certainly makes the code more Alpine-y - it doesn't actually fix the problem. The field populates properly (which it also did before) but I still need to actually click into the field and delete a character to get the filter system to be aware of that value. I was gonna record a video to show you what happens, but that would expose my exact long/lat for all to see on the Internet so...

**Reply by ryangjchandler** [reply]
Feel free to DM me on Discord with a video, it might help!

**Reply by intrepidws** [reply]
OK, just did thanks.

**Comment by intrepidws** [comment, marked as answer]
With lots of help from @ryangjchandler (thanks Ryan!) I was able to get this working. The problem was that while Alpine was aware of the value that I was setting, Livewire was not. My final code:

```php
Forms\Components\TextInput::make('zip_code_or_coords')
    ->label('... of this Zip Code')
    ->extraInputAttributes([
        'x-init' => '$wire.set(\'tableFilters.within_range.zip_code_or_coords\', $store.geolocation.coords)',
    ]),
```

And the corresponding JS code is:

```javascript
Alpine.store('geolocation', {
    latitude: null,
    longitude: null,
    coords: null,
    init() {
        if (window.navigator.geolocation) {
            window.navigator.geolocation.getCurrentPosition(position => {
                this.latitude = position.coords.latitude;
                this.longitude = position.coords.longitude;
                this.coords = this.latitude.toFixed(7) + ', ' + this.longitude.toFixed(7);
            })
        }
    }
})
```

**Reply by ryangjchandler** [reply]
It's worth noting that the use of `Alpine.store` here is just to make it more Alpine-y. If you had a binding to this value, it'd be reactive unlike a global JavaScript variable.
